### PR TITLE
[mlir][LLVM] Add nsw and nuw flags to trunc

### PR DESCRIPTION
**Pull request by lfrenot**
This implementation is based on the one already existing for the binary operations.

If the nuw keyword is present, and any of the truncated bits are non-zero, the result is a poison value. If the nsw keyword is present, and any of the truncated bits are not the same as the top bit of the truncation result, the result is a poison value.

@zero9178, @gysit and @Dinistro, could you take a look?

### CODE REVIEW

**Review by zero9178 — APPROVED** [review]
LGTM!

**Comment by tobiasgrosser** [comment]
Thank you. @lfrenot does not have merge rights, so please merge when you are happy with this.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:518]:
```suggestion
  let assemblyFormat = "$arg custom<OverflowFlags>($overflowFlags) `` attr-dict `:` type($arg) `to` type($res)";
```
nit: Can you move the attribute dictionary after the overflow flags. I think this is how it is done for the other operations that take overflow flags. 

I am somewhat surprised you don't need to elide overflow flags from the attribute dictionary when printing using `custom<LLVMOpAttrs>(attr-dict)`. Maybe tablegen does elide the flag since it is passed to a custom print function.

**Review by gysit — APPROVED** [review]
LGTM modulo nit.

`custom<LLVMOpAttrs>(attr-dict)` is apparently not necessary when looking at your tests.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:518]: Since the integer overflow flags are properties they should not be printed as part of attr-dict and `custom<LLVMOpAttrs>(attr-dict)` is not needed. In fact, it seems like the print/parseLLVMOpAttrs only makes sense for the fast math flags these days.

Comment (lfrenot) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:518]: I'm not sure what you mean by "`custom<LLVMOpAttrs>(attr-dict)` is not needed". Does anything need to be changed besides the nit above?

**Comment by Dinistro** [comment]
Side note: @lfrenot I suggest to directly request review from us instead of pinging, this ensures that stuff shows up in the correct places in GitHub 🙂

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:518]: Changing the nit should be sufficient (especially since the tests pass). The purpose of `custom<LLVMOpAttrs>(attr-dict)` is to filter out specific attributes from the attribute dictionary. This is not needed anymore now that the flag is a property.

**Comment by lfrenot** [comment]
> Side note: @lfrenot I suggest to directly request review from us instead of pinging, this ensures that stuff shows up in the correct places in GitHub 🙂

I would, but I do not have the perms to request reviews

**Comment by Dinistro** [comment]
> > Side note: @lfrenot I suggest to directly request review from us instead of pinging, this ensures that stuff shows up in the correct places in GitHub 🙂
> 
> I would, but I do not have the perms to request reviews

That seems weird, considering that you can ping people. I'll ask in the discord to figure out if this is intentional.

**Comment by tobiasgrosser** [comment]
@lfrenot may need commit access to get these rights. Lets give it a bit, before we ask chris for commit access.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (lfrenot) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td]: That worked, thank you

**Comment by Dinistro** [comment]
> @lfrenot may need commit access to get these rights. Lets give it a bit, before we ask chris for commit access.

This seems to be indeed be a GitHub restriction. Not sure why this is the case but that is not relevant for this PR.

**Comment by Dinistro** [comment]
I'm wondering if we should now also add these flags to `arith.trunci`?

**Pull request merged**
